### ISSUE TITLE
[Epic] APIKeys: Sunsetting of API keys 🌤️ 

### ISSUE THREAD

**Issue by eleijonmarck** [grafana/grafana#53567]
## Deprecation of API keys

**Context**

We are sunsetting API keys.

**Timeline**
Below is a detailed timeline for the deprecation of API keys in Grafana: 

- **By Aug. 31, 2024:** Creation of new API Keys will no longer be possible.
- **By Nov. 30, 2024:** Developers relying on provisioned API keys via automation should migrate to SATs using the guidelines provided as of Grafana 9.1.
- **By Jan. 31, 2025:** We will migrate any remaining API Keys to SATs automatically and remove the API key endpoints. Any automation relying on API Key provisioning will stop working.

In Grafana API keys have long been used for machine authentication, they have served users well until now.  🙏 We introduced service accounts and service account tokens (SAT) in Grafana 9.1 to replace API keys for machine authentication. All API keys present today in your Grafana instance are not affected until sunsetting, when we will migrate them over to SATs (all authentication for your application will remain consistent). 

**Why**
Service accounts allow you to rotate tokens without re-assigning permissions, mimic user behavior and work with role-based access control. This enables users to issue a service account alongside a service or a team within Grafana to manage their own set of keys. This also allows us to enhance Grafana with the additional feature of personal access tokens (PAT).
See more on the benefits of using service accounts over API keys - https://grafana.com/docs/grafana/latest/administration/service-accounts/#service-account-benefits

Or see this video overview:
https://drive.google.com/file/d/1lcF0KkZh9_4vrWBGgzpB8jFLwGGFJWxI/view?usp=sharing

For more information [Blog post](https://grafana.com/blog/2024/06/12/grafana-update-service-account-tokens-are-replacing-api-keys/).

## delivery plan
- [x] Grafana 9.1: Introduce service accounts in the org admin section, [announce API keys deprecation](https://grafana.com/blog/2022/08/24/new-in-grafana-9.1-service-accounts-are-now-ga/)
- [x] [Deprecation notice for API keys](https://github.com/grafana/grafana/pull/64436) endpoints (API calls and terraform API calls)
- [x] [Blog post](https://grafana.com/blog/2024/06/12/grafana-update-service-account-tokens-are-replacing-api-keys/) announcement for deprecation period and strategies
- [ ] Remove/Disable the endpoints for API keys to only allow the created API keys at that time to be functioning
- [ ] Remove API key endpoints entirely

#### *API*

from 
```
POST /api/apikeys
```

is replaced by

```
POST /api/serviceaccount
POST /api/serviceaccount/<id>/token
```

#### *Terraform*
from
```tf
resource "grafana_api_key" "api_key_name" {
     name = "api_key_name"
     
}
```

is replaced by
```tf
resource "grafana_service_account" "service_account_name" {
     name = "service_account_name"
     ..
}
resource "grafana_service_account_token" "service_account_token_name" {
     name = "service_account_token_name"
     ..
}

```

**Comment by ttorcz** [comment]
Could you add ability to _provision_ Service Account?

**Comment by dnck** [comment]
Just an FYI @eleijonmarck 

Running a test with image, `docker.io/grafana/grafana:9.3.1-ubuntu`, and I see the following logline on start-up:
```
WARN [12-05|21:12:21] secret scan interval is too low, increasing to 5m0s logger=serviceaccounts.background
```
I've tried various ways to get rid of that warning, e.g. setting `GF_SERVICE_ACCOUNTS_ENABLED="false"`,   `GF_SERVICE_ACCOUNTS_SECRET_SCAN_INTERNAL="5m0s"` to no avail. I've also tried logging debug from the `serviceaccounts.background` filter, etc. and I see nothing to indicate where I could set this variable.

**Comment by eleijonmarck** [comment]
@dnck hey dan thanks for asking here. If I understand correctly , you would like to disable or change the scan interval of secret scanner for the service accounts.
I have added a new issue for this discussion as it is not related to sunsetting apikeys. https://github.com/grafana/grafana/issues/59880

fyi @Jguer

**Comment by ttorcz** [comment]
> there is a functionality of provisioning ServiceAccounts via terraform registry

I was looking for provisioning of ServiceAccounts through config files, as it is possible now for datasources, dashboards, alerts etc. :
https://grafana.com/docs/grafana/latest/administration/provisioning/

**Comment by eleijonmarck** [comment]
@ttorcz for now we only support datasources, dashboards, alerts via provisioning files.
As of today we are not planning to provision service accounts through provisioning files.

To provision service accounts, it would be through terraform.

**Comment by eleijonmarck** [comment]
Parked until we pick this up again

Next step is to make sure we create serviceaccounttokens for each api key on the calls.

**Comment by kalleep** [comment]
Before we can disable api key endpoints we should review what internal services we have that would break and give them time to migrate to service accounts instead.

Right now I can see:
* Machine learning - pr is up https://github.com/grafana/machine-learning/pull/4001
* Incident - Issue to use service accounts instead of api keys https://github.com/grafana/incident/issues/2081

**Comment by mcschatz** [comment]
@kalleep We discussed this in planning this week. @twu is taking lead to make this migration happen. Thank you for ensuring we're set before disabling. Sorry for blocking!

**Comment by twu** [comment]
@eleijonmarck We released Incident `v1.48.0` (2024-10-16) today which contains support for Service Accounts (grafana/incident#5237). This means Incident should no longer be a blocker for this to be merged 🤞 🙇

**Comment by eleijonmarck** [comment]
Added a task to remove from terraform-provider test - https://github.com/grafana/terraform-provider-grafana/issues/1791

**Comment by kalleep** [comment]
@dmihai should we keep this one open until we have done clean-up in the code?
